### PR TITLE
fix(frontend): create two tests for the unauthorised component

### PR DESCRIPTION
**Pull request by Slav666**

### CODE REVIEW

File: client/src/app/unauthorized.component.test.tsx [hunk 1]
```diff
@@ -0,0 +1,24 @@
...
+
+import { render, screen, userEvent, waitFor } from '~/test/utils';
+
+import { Unauthorized } from './unauthorized.component';
+
+describe('Unauthorized', () => {
+  it('should render unauthorized component', () => {
+    render(<Unauthorized />);
+
+    expect(screen.getByRole('heading', { name: /Unauthorized/i })).toBeInTheDocument();
```
Comment (marksmall) [inline client/src/app/unauthorized.component.test.tsx]: by using the `//i` syntax, which means regex case agnostic, then you shouldn't capitalize anything, keep it all lowercase. That is how Kent Dodds and other bloggers do it.

File: client/src/app/unauthorized.component.test.tsx [hunk 2]
```diff
@@ -0,0 +1,24 @@
...
+    expect(screen.getByRole('button', { name: /Go Back/i })).toBeInTheDocument();
+  });
+
+  it('should move one step backward from unauthorized page when clicks on the "Go Back" button', async () => {
+    const goBack = vi.fn();
+    render(<Unauthorized goBack={goBack} />);
+
+    userEvent.click(screen.getByRole('button', { name: 'Go Back' }));
+
+    await waitFor(() => expect(goBack).toBeCalled());
```
Comment (marksmall) [inline client/src/app/unauthorized.component.test.tsx]: I think `toBeCalled` is an alias, we have never used this in any of our code, always use `toHaveBeenCalled`. If you do find other code doing this, then flag it for change, as it is wrong. I have no added an eslint rule that catches this, so you will see an error in future.

File: client/src/app/unauthorized.component.tsx [hunk 3]
```diff
@@ -1,12 +1,10 @@
...
-
-export const Unauthorized: FC = (): ReactElement => {
-  const navigate = useNavigate();
-
-  const goBack = () => navigate(-1);
+interface Props {
+  goBack?: () => void;
+}
 
+export const Unauthorized: FC<Props> = ({ goBack }): ReactElement => {
```
Comment (marksmall) [inline client/src/app/unauthorized.component.tsx]: You can now make this an `implicit` return, rather than `explicit`. You should only use explicit returns, when you have functions/hooks etc, like you did in the previous code. This is another pattern I want to enforce, so  our code is as similar as possible, making everyone's lives easier.

**Review by marksmall — CHANGES_REQUESTED** [review]
Couple of comments, but a far better PR than the previous one.

**Comment by marksmall** [comment]
Instead of making commits of one line saying `add requested changes....` which is really cryptic. If I go back to the commits in 3 weeks time, how do I find the PR that the commit relates to, where there is the conversation? Every request for change should be their own `fix` commit. The `title` should be a short version of what was asked to be changed and the body, hopefully should be able to get that from the PR conversation and then what you did and why. Single line commits are not acceptable, I will just delete code that does this. If you wrote the code, you must be able to explain it in plain english. I know English is a bad language, maybe write it in Polish and translate it to English, Google Translate is pretty good. Basically, it's up to you to explain what you did and why, not rely on others figuring it out.

File: .vscode/settig.json
Comment (marksmall) [inline .vscode/settig.json:1]: The name of this file is wrong and will not work, simple typo, it should be `settings.json`. I've not gon through this file, but I hope you got it from my `orbison` code, as there will have been changes. If you have just used an old one from say the blogs app, it could easily be missing stuff.

File: client/src/app/unauthorized.component.test.tsx [hunk 4]
```diff
@@ -0,0 +1,24 @@
...
+    expect(screen.getByRole('heading', { name: /unauthorized/i })).toBeInTheDocument();
+    expect(screen.getByText(/you do not have access to the requested page./i)).toBeInTheDocument();
+    expect(screen.getByRole('button', { name: /go back/i })).toBeInTheDocument();
+  });
+
+  it('should move one step backward from unauthorized page when clicks on the "Go Back" button', async () => {
+    const goBack = vi.fn();
+    render(<Unauthorized goBack={goBack} />);
+
+    userEvent.click(screen.getByRole('button', { name: 'Go Back' }));
```
Comment (marksmall) [inline client/src/app/unauthorized.component.test.tsx]: I thought I had previously linked to how to use the userEvent, please use it how the docs use it, not in some random stackoverflow, I got caught out by doing that. See https://testing-library.com/docs/user-event/intro/, it has an example. So there is setting up the user, then using the user to click things, it should be pretty self-explanatory, but if not, just ask.

One of the things is to await the click, that should probably negate the expect also having to be await'd, making it more linear, async stuff is hard.

**Review by marksmall — CHANGES_REQUESTED** [review]
Sorry, couple of more things, but what your producing is good, just a tad outdated, by like 6 months.... Software moves fast.

File: client/src/app/unauthorized.component.test.tsx [hunk 5]
```diff
@@ -0,0 +1,25 @@
...
+  it('should render unauthorized component', () => {
+    render(<Unauthorized />);
+
+    expect(screen.getByRole('heading', { name: /unauthorized/i })).toBeInTheDocument();
+    expect(screen.getByText(/you do not have access to the requested page./i)).toBeInTheDocument();
+    expect(screen.getByRole('button', { name: /go back/i })).toBeInTheDocument();
+  });
+
+  it('should move one step backward from unauthorized page when clicks on the "Go Back" button', async () => {
+    const user = userEvent.setup();
```
Comment (RoryMacGregor88) [inline client/src/app/unauthorized.component.test.tsx:19]: I'm not familiar with this use of `userEvent`, what is this doing?

File: client/src/app/unauthorized.component.test.tsx [hunk 6]
```diff
@@ -0,0 +1,25 @@
...
+    const goBack = vi.fn();
+    render(<Unauthorized goBack={goBack} />);
+
+    await user.click(screen.getByRole('button', { name: /go back/i }));
```
Comment (RoryMacGregor88) [inline client/src/app/unauthorized.component.test.tsx:22]: I've stopped using regex to match things in tests, because it's not testing for text that should be capitalised. Like if it's supposed to be `Go Back` but I type `gO bACk` it would still pass.

File: client/src/app/unauthorized.component.tsx [hunk 7]
```diff
@@ -1,25 +1,21 @@
 import { FC, ReactElement } from 'react';
 
-import { useNavigate } from 'react-router-dom';
+interface Props {
+  goBack?: () => void;
```
Comment (RoryMacGregor88) [inline client/src/app/unauthorized.component.tsx]: This shouldn't be optional, as it is always being passed.

File: client/src/app/unauthorized.component.test.tsx (hunk 6 above)
Comment (stevefaeembra) [reply client/src/app/unauthorized.component.test.tsx:22]: the /i makes this case insensitive, good practice is to lowercase search term and add this

File: client/src/app/unauthorized.component.test.tsx (hunk 5 above)
Comment (Slav666) [reply client/src/app/unauthorized.component.test.tsx:19]: There is a link to the documentation which explains that: https://testing-library.com/docs/user-event/intro/ Mark  just asked me to do it in that way.

File: client/src/app/unauthorized.component.tsx (hunk 7 above)
Comment (Slav666) [reply client/src/app/unauthorized.component.tsx]: This is optional because the first test doesn't need the goBack props. Do I need to refactor the first test with goBack props?

File: client/src/app/unauthorized.component.test.tsx (hunk 5 above)
Comment (RoryMacGregor88) [reply client/src/app/unauthorized.component.test.tsx:19]: Ok cool, good to know

File: client/src/app/unauthorized.component.tsx (hunk 7 above)
Comment (RoryMacGregor88) [reply client/src/app/unauthorized.component.tsx]: Yeah I'd pass it in every test. I used to only pass the props needed for each test, but I realised that having missing props in tests could lead to unexpected behaviour, so now I add all props in all tests.

File: client/src/app/unauthorized.component.test.tsx (hunk 6 above)
Comment (RoryMacGregor88) [reply client/src/app/unauthorized.component.test.tsx:22]: For things like search terms or forms, where we'd be lowercasing everything anyway I imagine it would be redundant to make matchers case-insensitive.

But for things like buttons, typos could slip through. Like if I typed `go Back` as the button text, this test wouldn't catch the mistake because of the case-insensitive regex.

Comment (marksmall) [reply client/src/app/unauthorized.component.test.tsx:22]: I have yet to see an example by the creators of testing-library use plain strings. I don't think unit/integration testing should care about typos like this. That would more likely be the place to test using **snapshots** or E2E testing, or even **stories**. Use the right kind of test to test what is appropriate. That being said, I originally thought the same as Rory, but I seem to be in the minority, if you look at blogs|tutorials|the docs etc.

File: client/src/app/unauthorized.component.tsx (hunk 7 above)
Comment (marksmall) [reply client/src/app/unauthorized.component.tsx]: The test shouldn't drive the component, only things that are optional during actual live running should be optional. 

Remember when developing, interfaces|types tell the developer what they need to pass during **actual usage**, not during testing, where you may only be focussing on a part of the functionality. So, while the one single test might not need it to work, due to your focus being only on what DOM elements are present. The component would be broken, as when the user clicked on it, it would do nothing. It is also possible, as Rory states, that not providing the prop, could have some consequences you didn't foresee.

File: client/src/app/unauthorized.component.test.tsx (hunk 5 above)
Comment (marksmall) [reply client/src/app/unauthorized.component.test.tsx:19]: This is not something I am asking anyone to do, this is what the docs say, don't confuse the two. How to use `userEvent` has changed fairly recently, so everyone has to update their knowledge on how to use it.

File: client/src/app/unauthorized.component.test.tsx (hunk 6 above)
Comment (marksmall) [reply client/src/app/unauthorized.component.test.tsx:22]: What matters most is we are consistent. I don't want some using plain strings and others using regex, do what the docs suggets, don't make it up as you go along @RoryMacGregor88. If you disagree, fine, state reasons, argue over it and agree, but don't arbitrarily, do things different just cause you disagree.
